kernel/kernel-api/src/test/scala/io/delta/kernel/internal/TableConfigSuite.scala

```diff
@@ -0,0 +1,163 @@
+    }
+    assert(e1.getMessage.contains("""Unable to parse {"key": "value"}"""))
+
+    val engine2 = mockEngine(jsonHandler = new ReturnNullRowsJsonHandler(2))
+    val e2 = intercept[IllegalArgumentException] {
+      TableConfig.parseJSONKeyValueMap(engine2, """{"key": "value"}""")
+    }
+    assert(e2.getMessage.contains("Iterator contains more than one element"))
+
+    val errMsg = "Close called failed"
+    val engine3 = mockEngine(jsonHandler = new ReturnCloseFailIteratorJsonHandler(errMsg))
+    val e3 = intercept[KernelException] {
+      TableConfig.parseJSONKeyValueMap(engine3, """{"key": "value"}""")
+    }
+    assert(e3.getMessage.contains(s"java.io.IOException: $errMsg"))
+
+    val engine4 = mockEngine(jsonHandler = new ReturnNullRowsJsonHandler(1))
+    val e4 = intercept[IllegalArgumentException] {
+      TableConfig.parseJSONKeyValueMap(engine4, """{"key": "value"}""")
+    }
+    assert(e4.getMessage === null)
+  }
+}
+
+/**
+ * Mock JsonHandler which returns a ColumnarBatch with a single row containing a Map[String, String]
+ * column.
+ */
+class KeyValueJsonHandler(map: Map[String, String]) extends BaseMockJsonHandler {
+  val expSchema: StructType =
+    new StructType().add("config", new MapType(StringType.STRING, StringType.STRING, true))
+  override def parseJson(
+    jsonStringVector: ColumnVector,
+    outputSchema: StructType,
+    selectionVector: Optional[ColumnVector]): ColumnarBatch = {
+    assert(outputSchema == expSchema)
+    new ColumnarBatch {
+      override def getSchema: StructType = outputSchema
+
+      override def getColumnVector(ordinal: Int): ColumnVector = {
+        ordinal match {
+          case 0 => mapTypeVector(map)
+        }
+      }
+
+      override def getSize: Int = 1
+    }
+  }
+
+  def mapTypeVector(map: Map[String, String]): ColumnVector = {
+    new ColumnVector {
+      override def getDataType: DataType = new MapType(StringType.STRING, StringType.STRING, true)
+
+      override def getSize: Int = 1
+
+      override def close(): Unit = {}
+
+      override def isNullAt(rowId: Int): Boolean = rowId >= 1
+
+      override def getMap(rowId: Int): MapValue = VectorUtils.stringStringMapValue(map.asJava)
+    }
+  }
+}
+
+/**
+ * Mock JsonHandler which returns a ColumnarBatch with nRow rows containing a single null column.
+ */
+class ReturnNullRowsJsonHandler(nRow: Int) extends BaseMockJsonHandler {
+  override def parseJson(
+    jsonStringVector: ColumnVector,
+    outputSchema: StructType,
+    selectionVector: Optional[ColumnVector]): ColumnarBatch = {
+    new ColumnarBatch {
+      override def getSchema: StructType = outputSchema
+
+      override def getColumnVector(ordinal: Int): ColumnVector = {
+        stringVector(List.fill(nRow)(null).asInstanceOf[List[String]].asJava)
+      }
+
+      override def getSize: Int = nRow
+    }
+  }
+}
+
+/**
+ * Mock JsonHandler which returns a ColumnarBatch with an iterator that throws an IOException when
+ * close is called.
+ */
```

Comment:
should this be throwing error when parseJson is called instead?

Reply:
This error will be thrown in `getSingularElement`. According to the `parseJson` API, I think it will not throw `IOException`?